channelserver/src/meta.rs

```diff
@@ -0,0 +1,219 @@
+}
+
+impl Default for SenderData {
+    fn default() -> Self {
+        Self {
+            ua: None,
+            addr: None,
+            city: None,
+            region: None,
+            country: None,
+        }
+    }
+}
+
+// Parse the Accept-Language header to get the list of preferred languages.
+// We default to "en" because of well-established Anglo-biases.
+fn preferred_language(alheader: String) -> Vec<String> {
+    let default_lang = String::from("en");
+    let mut lang_tree: BTreeMap<String, String> = BTreeMap::new();
+    let mut i = 0;
+    alheader.split(",").for_each(|l| {
+        if l.contains(";") {
+            let weight: Vec<&str> = l.split(";").collect();
+            let lang = weight[0].to_ascii_lowercase();
+            let pref = weight[1].to_ascii_lowercase();
+            lang_tree.insert(String::from(pref), String::from(lang));
+        } else {
+            lang_tree.insert(
+                format!("q=1.{:02}", i),
+                String::from(l.to_ascii_lowercase()),
+            );
+            i += 1;
+        }
+    });
+    let mut langs: Vec<String> = lang_tree.values().map(|l| l.to_owned()).collect();
+    langs.reverse();
+    langs.push(default_lang);
+    langs
+}
+
+// Return the element that most closely matches the preferred language.
+// This rounds up from the dialect if possible.
+fn get_preferred_language_element(
+    langs: &Vec<String>,
+    elements: BTreeMap<String, String>,
+) -> Option<String> {
+    for lang in langs.clone() {
+        if elements.contains_key(&lang) {
+            //unwrap is safe-ish because we've checked that the key exists.
+            return Some(elements.get(lang.as_str()).unwrap().to_string());
+        }
+        if lang.contains("-") {
+            let (lang, _) = lang.split_at(2);
+            if elements.contains_key(lang) {
+                return Some(elements.get(lang).unwrap().to_string());
+            }
+        }
+    }
+    None
+}
+
+// Set the sender meta information from the request headers.
+impl From<HttpRequest<WsChannelSessionState>> for SenderData {
+    fn from(req: HttpRequest<WsChannelSessionState>) -> Self {
+        let mut sender = SenderData::default();
+        let headers = req.headers();
+        //TODO: Get the default lang
+        let langs = match headers.get(http::header::ACCEPT_LANGUAGE) {
+            None => vec![String::from("en")],
+            Some(l) => preferred_language(l.to_str().unwrap_or("").to_owned()),
+        };
+        let conn = req.connection_info();
+        // parse user-header for platform info
+        sender.ua = match headers.get(http::header::USER_AGENT) {
+            None => None,
+            Some(s) => match s.to_str() {
+                Err(x) => {
+                    println!("Error: bad UA string: {:?}", x);
+                    None
+                }
+                Ok(s) => Some(s.to_owned()),
+            },
+        };
+        sender.addr = match conn.remote() {
```

Comment:
The docs say that this method traverses `X-Forwarded-For` and similar headers.  Are we able to control the depth of that traversal in order to prevent IP spoofing via fake X-F-F header?  It doesn't look so at a glance:

  https://actix.rs/actix-web/src/actix_web/info.rs.html#98

We'll need some story about how to prevent spoofing here for the metadata to be useful.

Reply:
Good question. Let me dig into that a bit.

Reply:
Hmm. looks like it just returns the first value. If we want to do something better, we'll have to parse and handle it ourselves. 

I'm ok with doing that, but I think that should be a separate issue.

Reply:
Taking the first entry here might just give us e.g. the local IP of nginx, or the IP of the loadbalancer, so I think we'll definitely have to do *something* here.  I guess it's possible that that "something" could be in operational config though.  How do we expect to run this in production, fronted by what services?

I'm fine with a follow-up issue, but we definitely can't lose track of it.